### PR TITLE
Add TIBX archive parser

### LINKED ISSUE #20: Add support for TIB and TIBX format (Acronis backup)
We came across the tib/tibx format (see [https://kb.acronis.com/content/63441).](https://kb.acronis.com/content/63441).)

We should add support for it

### PR DESCRIPTION
**Pull request by falklindner**
Add a parser for the Acronis CyberBackup/CyberProtect TIBX ("archive3") backup format as a `dissect.archive.tibx` subpackage.

Closes #20 

## Proposed changes

The parser handles the full TIBX format:

- **Page store**: 4 KiB pages with CRC-32C verification, ARCH commit-root superblocks, live-root / commit-root selection
- **LSM B-tree index**: TLV directory, L-SB superblocks, LEAF/LDIR cell decoding, data_map / segment_map with mem-tree and on-disk ctree layers, tombstone masking for backup-version cleanup
- **Compression**: zstd and linked-LZ4 (dictionary-seeded block chain)
- **Encryption**: AES-256-CBC with PBKDF2 key derivation
- **Chain reconstruction**: full incremental/differential chain with recovery-point selection
- **Volume reconstruction**: NTFS, FAT32, exFAT, FAT12/16 and ext boot-sector detection for accurate volume sizing
- **Archive variants**: split archives (auto-stitched), version sets (TLV[18] file table with compacted bases), multi-disk archives

TIBX is materially more complex than VBK (page store + LSM index + crypto + chain reconstruction), so a subpackage is proposed rather than a flat module. Happy to consolidate if preferred.

### Dependencies

`pycryptodome` and `backports.zstd` added to the `full` optional extra; `lz4` added to the test dependency group.

### Testing

70 unit tests covering all parser layers (page store, LSM trees, segments, codecs, crypto, volume reconstruction, version sets, tombstone masking). Tests are fully hermetic — they build synthetic CRC-correct archives via `tests/_synth.py`, no binary fixtures required.

Validated byte-for-byte against real Acronis True Image 2026 (Build 42980) output across: single-partition, multi-partition, whole-disk (MBR and GPT), multi-disk, split, incremental, differential, and encrypted archives; NTFS, FAT32, and exFAT volumes.

### Acknowledgments

Thanks to @mniedermaier for his contributions and support in developing this parser. 

## Checklist

- [x] PR title is descriptive of the changes
- [x] The description is descriptive of the changes
- [x] Tests are included and pass
- [x] Closes #20

### CODE REVIEW

**Comment by yunzheng** [comment]
Very nice work, and thanks for sharing it! I haven't started reviewing yet, but I pushed a commit filling in a few of the unknown fields based on some of our own reversing efforts.

**Comment by yunzheng** [comment]
Regarding the last commit, can the TIBX encryption not be detected via the `encr_alg` in the superblock header?

**Comment by falklindner** [comment]
> Regarding the last commit, can the TIBX encryption not be detected via the `encr_alg` in the superblock header?

Good point — checked against the full real-sample corpus (12 archives): encr_alg in the ARCH superblock agrees with the keymap detection everywhere (AES_256_CBC on the encrypted sample, NONE on the rest), so yes, it does indicate encryption.

The keymap scan answers a narrower question though: not "is this encrypted" but "does a password-wrapped data key exist" — which is what drives the password prompt / keychain lookup. Acronis also supports certificate-wrapped keys (format byte 0x02 next to the 0x01 we locate), where encr_alg would be set but asking for a password can't help. Detection via the keymap also can't disagree with unlock(), since both parse the same structure, and it doesn't depend on the superblock byte layout, which we've only verified for header_version 8.

The two compose nicely, and encr_alg is essentially free: happy to use encr_alg != NONE as the archive-level signal and the keymap scan to decide password vs. certificate — that would let a certificate-wrapped archive fail with "encrypted, but not password-protected" instead of a generic missing-password error.

**Comment by falklindner** [comment]
Is there anything I can do to support the PR?

File: dissect/archive/tibx/c_tibx.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+``0x41`` (``'A'``), byte 1 the page type, and a CRC-32C of the page (checksum bytes
+zeroed) is stored at ``+0x04``. The page content ("body") starts after this 8-byte
+envelope.
+
+Most multi-byte fields are big-endian, so the definitions are loaded into a big-endian
+cstruct instance. The few little-endian fields (the LSM cell-group header, the LEAF
+sequence id and the segment_map ``page_count``) are decoded manually at their use sites.
+
+References:
+    - acronis-tibx ``docs/FINDINGS.md`` (reverse-engineered format specification)
```
Comment (yunzheng) [inline dissect/archive/tibx/c_tibx.py]: This file seems to be missing in the PR.

File: dissect/archive/tibx/crc32c.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+"""CRC-32C (Castagnoli) for TIBX page verification.
+
+Every 4 KiB TIBX page stores a CRC-32C of itself big-endian at offset ``0x04``, computed
+with the four checksum bytes zeroed.
+
+The table implementation is ported from the MIT-licensed ``acronis-tib-reader`` project
+(see THIRD_PARTY_NOTICES.md).
```
Comment (yunzheng) [inline dissect/archive/tibx/crc32c.py]: This seems to be missing. Would be nice to credit where credit is due, either in the PR or as links in the code.

File: dissect/archive/tibx/crypto.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    from Crypto.Protocol.KDF import PBKDF2
+
+    alg = blob[offset + 1]
+    iter_log2 = blob[offset + 2]
+    if alg in GCM_ALG_IDS:
+        raise UnsupportedFormatError("AES-GCM encrypted TIBX archives are not supported")
+    if not _blob_is_well_formed(blob, offset, cbc_only=True):
+        return None
+    salt = blob[offset + 4 : offset + 4 + SALT_SIZE]
+    wrapped = blob[offset + 4 + SALT_SIZE :]
```
Comment (yunzheng) [inline dissect/archive/tibx/crypto.py]: Is it possible to turn this into a readable structure instead of using offsets?

File: dissect/archive/tibx/lsm.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
+"""LSM index layer: the ARCH header's TLV directory, L-SB superblocks and tree walking.
+
+The ARCH header body carries a 19-slot TLV directory (at body offset ``0x400``); slots
+0-8 hold **L-SB** superblocks, one per LSM tree (slot 1 = ``data_map``, slot 2 =
+``segment_map``, slot 5 = ``slices``, slot 7 = ``keymap``, slot 18 = volume table).
+Each L-SB describes the tree's on-disk ctree runs (B-tree roots of LEAF/LDIR pages)
+plus a residual **mem-tree** holding not-yet-compacted records -- small archives keep
+*all* records in the mem-tree with empty ctrees.
+
+Ported from the MIT-licensed acronis-tib-reader engine (see THIRD_PARTY_NOTICES.md).
```
Comment (yunzheng) [inline dissect/archive/tibx/lsm.py]: missing, either add the the file or fix the text inline by adding the references.

File: dissect/archive/tibx/page.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        return c_tibx.PageType(value).name
+    except ValueError:
+        return f"{value:#04x}"
+
+
+class SuperBlock:
+    """An ARCH superblock (commit root) at a given byte offset in the page store."""
+
+    def __init__(self, page: bytes, offset: int):
+        if page[8:12] != ARCH_MAGIC:
```
Comment (yunzheng) [inline dissect/archive/tibx/page.py]: While the `Page` structure is really simple, I wonder if generally we should use a structure where possible or just use offsetting on bytes. The latter is of course faster, but I somehow prefer to work on structs.

From our own reversing efforts we used this structure for the Page:

```
struct Page {
    uint32 magic;
    uint32 crc32;
    uint8 data[PAGE_SIZE-8];
};
```

Paging (pun intended) other dissect maintainers (especially @Schamper, @twiggler, others feel free to jump in) on what should be generally be used as best practice in the overall dissect code base.

File: tests/_synth.py [hunk 6]
```diff
@@ -0,0 +1,604 @@
+"""Synthetic, CRC-correct TIBX archive builders for hermetic tests.
+
+Real archives can't be committed for every edge case (and corruption cases can't be
+produced by Acronis at all), so structural tests build minimal valid page stores from
+scratch — up to and including complete archives with a TLV directory, data_map /
+segment_map LSM superblocks (mem-tree or on-disk LEAF/LDIR ctrees) and SG data
+segments, which the acronis-tibx POC parses as well (the oracle in spikes/).
+
+Ported from the acronis-tibx POC's synthetic fixtures (MIT, see THIRD_PARTY_NOTICES.md)
+and extended with the LSM layer.
```
Comment (yunzheng) [inline tests/_synth.py]: Missing references, what is acronis-tibx, and where can we find it. Either add the file or fix the docstring by adding the references inline.

File: dissect/archive/tibx/crypto.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
+"""Encrypted-archive support: recover the data key from a password, decrypt segments.
+
+The LSM index is always plaintext; only data segments are encrypted (page tag ``SE``).
+The data key is wrapped with a password-derived key-encryption key (KEK) and stored in
+the keymap tree (TLV[7]) superblock mem-tree. A wrapped-key blob is:
+
+    [format=0x01][alg][iter_log2][reserved][salt||16][wrapped key, PKCS#7-padded||16n]
+
+    KEK      = PBKDF2-HMAC-SHA256(password, salt, 1 << iter_log2, 32 bytes)
+    data key = PKCS#7-unpad(AES-256-CBC-decrypt(wrapped, KEK, IV=0))
```
Comment (yunzheng) [inline dissect/archive/tibx/crypto.py]: Can this wrapped encryption key be turned into a structure? see my other comments about offsetting using hardcoded numbers.

**Review by yunzheng — COMMENTED** [review]
Hi @falklindner, sorry for the late reply, but I was on holiday.

I left some general feedback.

Also, regarding the synthetic tests, while I do think that's awesome, I think nothing beats real data. It would be nice to add tibx archives created by Acronis itself, for example:

* simple tibx archive
* password protected archive
* full disk vs files
* other edge cases that we want to test

I'm also curious how much you're already using this implementation for your own needs.

File: dissect/archive/tibx/c_tibx.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+from __future__ import annotations
+
+from dissect.cstruct import cstruct
+
+tibx_def = """
+#define PAGE_SIZE 0x1000
+
+enum PageType : uint8 {
+    ARCH    = 0x01,     /* superblock / commit root (+ continuation pages) */
+    ARCI    = 0x02,
```
Comment (yunzheng) [inline dissect/archive/tibx/c_tibx.py]:
```suggestion
    ARCI    = 0x02,     /* commit info */
```

File: dissect/archive/tibx/crypto.py (hunk 3 above)
Comment (falklindner) [reply dissect/archive/tibx/crypto.py]: Done — added a wrapped_key struct in c_tibx.py; crypto.py parses with it instead of offset arithmetic.

File: dissect/archive/tibx/c_tibx.py (hunk 8 above)
Comment (falklindner) [reply dissect/archive/tibx/c_tibx.py]: Applied, thanks.

File: dissect/archive/tibx/page.py (hunk 5 above)
Comment (falklindner) [reply dissect/archive/tibx/page.py]: Numbers from this codebase, per page envelope: c_tibx.page_header(page) 3.99 µs vs direct byte indexing 0.24 µs (16.7×). Only matters in the three whole-archive scan loops (live_root, commit_roots, verify) — on a 100 GiB archive, ~100 s of header parsing vs ~6 s. Current state is a hybrid: structures everywhere except those loops, reasoning noted at the top of page.py. Happy to flip to structures everywhere if uniformity is preferred. Your struct Page matches what we have.

File: tests/_synth.py (hunk 6 above)
Comment (falklindner) [reply tests/_synth.py]: Answered in THIRD_PARTY_NOTICES.md (93dc6f7).

acronis-tibx is an MIT research project by mniedermaier (Copyright (c) 2026) that is no longer publicly available — the synthetic fixtures here derive from its container-level test builders and are extended with the LSM layer. Its own LSM engine was vendored from acronis-tib-reader, which is still public: https://github.com/TreadingTheTiber/acronis-tib-reader

See my reply on lsm.py — the attribution needed correcting, not just a reference.

File: dissect/archive/tibx/crypto.py (hunk 7 above)
Comment (falklindner) [reply dissect/archive/tibx/crypto.py]: Done — added a wrapped_key struct in c_tibx.py; crypto.py parses with it instead of offset arithmetic.

File: dissect/archive/tibx/codec.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+# Generous ceiling for a single decompressed unit; real segments are a few MiB
+MAX_DECOMPRESSED = 2 << 30
+
+# Optional C-accelerated LZ4 block decoder. The pure-Python decoder below is always the
+# fallback (``lz4`` is not a hard dependency), so nothing breaks without it; when present
+# it is orders of magnitude faster and, per a capability probe, produces byte-identical
+# output including the growing-dictionary (``LZ4_decompress_safe_continue``) semantics the
+# linked-LZ4 chains rely on.
+try:
+    import lz4.block as _lz4_block
```
Comment (Schamper) [inline dissect/archive/tibx/codec.py]: LZ4 is also available in `dissect.util` with a Rust powered backend if available, can you switch to that?

File: dissect/archive/tibx/exceptions.py
Comment (Schamper) [inline dissect/archive/tibx/exceptions.py]: Can you make file names singular? `exception.py`.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/archive/tibx/page.py (hunk 5 above)
Comment (yunzheng) [reply dissect/archive/tibx/page.py]: Thanks for benchmarking, that really helps with the decision IMO. I think the current hybrid tradeoff is good and documenting the reason helps as well.

marking this as resolved.

Comment (Schamper) [reply dissect/archive/tibx/page.py]: Disclaimer: I didn't look at the surrounding code very much, but just based on this comment chain, is the benchmark realistic? I imagine most use-cases for this in Dissect would involve random access and not a full linear sweep, if you combine that with a realistic LRU cache on the page itself, what does the real cost become? Not that I don't agree with the decision, but I do want it to be based on sound benchmarks.

Comment (falklindner) [reply dissect/archive/tibx/page.py]: You're right to push on this — I re-measured end to end instead of timing the parse call in isolation, and the 16.7x I quoted doesn't justify what I used it for. Two of the three loops didn't need the offsetting at all.

On a 949 MiB archive (242,977 pages), offset vs struct:

| loop | pages touched | offset | struct | ratio |
|---|---|---|---|---|
| `live_root()` — every open | **3 of 242,977** | 686 µs | 702 µs | **1.02x** |
| `commit_roots()` — recovery points | all | 859 ms | 1644 ms | **1.92x** |
| `verify()` | all | 13.1 s | 13.9 s | **1.06x** |
| random access, 200 × 4 KiB | — | 0.2 ms, already all-struct | | — |

Three things I had wrong. `live_root()` isn't a sweep — it scans *backward* from EOF and the newest ARCH is within the last few pages, so it exits after 3. `verify()` is ~99% CRC, so the envelope choice is invisible there. And on the one real sweep, I/O dominates: 778 ms of `read()` against 81 ms of offset-parsing, so the per-envelope ratio (~11x here) collapses to 1.9x end to end.

On the LRU specifically: it changes nothing, and the reason is worth stating — a linear sweep gets **0 hits out of 242,977 reads**, since every page is visited exactly once. A 512-page LRU is pure overhead there. It neither rescues the sweep nor hides anything in the numbers. And your instinct about random access is right: that path never touches this code, it's structs throughout.

So `live_root()` and `verify()` are back on `c_tibx.page_header`, and only `commit_roots()` still indexes directly, with the end-to-end figures in the comment instead of the per-call ratio. Happy to make that one uniform too — 1.9x on a once-per-archive operation is a defensible thing to spend for one less special case; your call.

Separately, chasing why `verify()` was taking 172 s turned up something worth flagging beyond this PR. `dissect.util.hash.__init__` rebinds its `crc32c` attribute to the native Rust implementation, but `from dissect.util.hash.crc32c import crc32c` resolves through `sys.modules` to the pure-Python submodule and bypasses the rebind — identical output, 5.7 MB/s instead of 80.9 MB/s. `verify()` on that archive went 172 s → 15.5 s from the import alone. Easy to land on by accident, so it might be worth an issue upstream: any project reaching for the submodule form silently gets the slow path.

File: dissect/archive/tibx/codec.py (hunk 9 above)
Comment (falklindner) [reply dissect/archive/tibx/codec.py]: Partly, and it's worth doing — but it can't be a straight swap, because `dissect.util`'s LZ4 has no dictionary parameter on either backend:

```python
def decompress(src, uncompressed_size: int = -1, return_bytearray: bool = False)
```

The LSM cell streams and mem-tree blobs are chains of LZ4 blocks with `LZ4_decompress_safe_continue` semantics — each block may back-reference previously decompressed output as its dictionary. I checked whether that's actually exercised or just theoretical, by decoding a corpus of 16 real Acronis archives twice, with and without the dictionary:

| | |
|---|---|
| chains | 307 |
| blocks | 1,238 |
| blocks with a non-empty dictionary | 931 |
| **of those, blocks whose output changes without it** | **931 — all of them** |
| longest chain / largest dictionary | 11 blocks / 11,347 bytes |

So the dictionary is load-bearing for 75% of blocks. Those can't move as things stand.

**The other 25% can, and should.** The first block of every chain, plus the plain-LZ4 segment path in `segment.py`, pass no dictionary. Against all 307 such real blocks `dissect.util` is byte-identical — 0 mismatches — and it bounds output and rejects truncated input the same way. Timings over that 1.13 MB block set:

| decoder | per pass |
|---|---|
| in-tree, with the optional `lz4` C package | 0.6 ms |
| **`dissect.util` (Rust)** | **8.2 ms** |
| `dissect.util` (pure Python) | 93 ms |
| in-tree pure Python | 116 ms |

The Rust backend is slower than the `lz4` package but **14x faster than the in-tree pure-Python fallback it would replace** — and that fallback is the code that actually runs whenever `lz4` isn't installed, since it's an optional dependency. So the dispatch becomes:

1. `lz4` C package if present (handles the dictionary, fastest)
2. otherwise, no dictionary → `dissect.util.compression.lz4`
3. otherwise → the in-tree decoder, now only for dictionary blocks without `lz4`

Two wrinkles worth naming. The in-tree decoder doesn't disappear — the dictionary path still needs it — so this buys fallback speed rather than fewer lines. And the two `dissect.util` backends raise different types: Rust raises `ValueError`, pure Python raises `CorruptDataError`, which is not a `ValueError`, so the call site needs `except (CorruptDataError, ValueError)` to normalise into `CorruptArchiveError`.

The clean end state is a `dictionary=`/prefix parameter upstream in `dissect.util` — a few lines on the Python side (seed the output buffer, strip the seed) and a real change on the Rust side. After that the in-tree decoder goes away entirely and this file is just framing plus zstd. Happy to open that PR against `dissect.util` if you'd take it; say the word and I'll do the swap here either way.

File: dissect/archive/tibx/exception.py
Comment (falklindner) [reply dissect/archive/tibx/exception.py]: Happy to — one question on scope first, because the singular form would make this file the odd one out rather than the conventional one. `exceptions.py` (plural) is what `dissect.archive`, `dissect.util` and `dissect.cstruct` all use today, and `dissect/archive/exceptions.py` sits one directory up from this file.

There's an option that sidesteps the naming question entirely: **drop `tibx/exceptions.py` and move the four classes into the existing `dissect/archive/exceptions.py`**. Every other parser in the package already imports from there — `vbk.py`, `vma.py` and `wim.py` all do — and tibx is the only one carrying its own module. That removes the plural filename without introducing a plural/singular split in the same package, and deletes a file rather than renaming one.

One detail if you like that: tibx has its own `Error` base, re-exported from `dissect/archive/tibx/__init__.py` and used as a catch-all in `tibx.py`. Folding it into the package `Error` would widen that catch to other parsers' exceptions, so I'd add a `TibxError(Error)` base for the four to subclass and keep the catch narrow.

Otherwise, if you'd rather have the rename: is the instruction meant for this file specifically, or for the package? Read broadly it also catches `codecs.py` → `codec.py` and `maps.py` → `map.py`. `codec.py` has an independent benefit — `codecs` currently shadows a stdlib module name.

Cost is 13 import sites for `exceptions` (7 source, 6 test), plus 5 and 4 for the other two. All mechanical; just want to do it once in the direction you want.

File: dissect/archive/tibx/exceptions.py
Comment (Schamper) [reply dissect/archive/tibx/exceptions.py]: No offense, but please be more concise. Plural names is what we used to have but we're moving towards singular names slowly to "fix" that up.

File: dissect/archive/tibx/page.py (hunk 5 above)
Comment (Schamper) [reply dissect/archive/tibx/page.py]: Tldr? A LRU cache can only aid random access, not a sweep.

File: dissect/archive/tibx/codecs.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+# Generous ceiling for a single decompressed unit; real segments are a few MiB
+MAX_DECOMPRESSED = 2 << 30
+
+# Optional C-accelerated LZ4 block decoder. The pure-Python decoder below is always the
+# fallback (``lz4`` is not a hard dependency), so nothing breaks without it; when present
+# it is orders of magnitude faster and, per a capability probe, produces byte-identical
+# output including the growing-dictionary (``LZ4_decompress_safe_continue``) semantics the
+# linked-LZ4 chains rely on.
+try:
+    import lz4.block as _lz4_block
```
Comment (Schamper) [reply dissect/archive/tibx/codecs.py]: Soooo, tldr? What's the conclusion?

File: dissect/archive/tibx/exception.py
Comment (falklindner) [reply dissect/archive/tibx/exception.py]: Sorry for lengthy text before, done in dabb829 — exception.py, codec.py, map.py. Left tests/test_exceptions.py and dissect/archive/exceptions.py alone since they predate this branch; happy to include them if you'd rather they came along.

File: dissect/archive/tibx/codec.py (hunk 9 above)
Comment (falklindner) [reply dissect/archive/tibx/codec.py]: Can't fully implement: dissect.util's LZ4 has no dictionary parameter. A full switch needs a dictionary= upstream; happy to PR that, then this file drops its own decoder. What is your opinion on that?

File: dissect/archive/tibx/page.py (hunk 5 above)
Comment (falklindner) [reply dissect/archive/tibx/page.py]: Yes, exactly. I measured: 0 cache hits across a 242,977-page sweep, since every page is visited once.

Result: `live_root()` and `verify()` are back on `c_tibx.page_header`  and only `commit_roots()` still indexes directly. Pushed in 7b495d0.

Comment (Schamper) [reply dissect/archive/tibx/page.py]: My question was if they LRU would help with a realistic random read scenario, not a full sweep.

Comment (falklindner) [reply dissect/archive/tibx/page.py]: Sorry, I misunderstood in regards to LRU. Measured it: 0 LRU hits on random reads either. Removed the LRU in 4873c7b.

File: dissect/archive/tibx/codec.py (hunk 9 above)
Comment (yunzheng) [reply dissect/archive/tibx/codec.py]: libarchive.so.3 uses `LZ4_decompress_safe` and the [function prototype](https://github.com/lz4/lz4/blob/0774d05537f9762f838f7ab541b7765f1a729cb5/lib/lz4.h#L208) is:

```c
LZ4LIB_API int LZ4_decompress_safe (const char* src, char* dst, int compressedSize, int dstCapacity);
```

So it doesn't have any dictionary argument. I cannot see any dictionary loading either, but maybe i'm missing something. @falklindner Is there something you can verify the correctness of the decompression?

Comment (falklindner) [reply dissect/archive/tibx/codec.py]: Good question — there's nothing to load. The "dictionary" is just the previous blocks' output, so a decoder writing successive blocks into one buffer gets those back-references for free. No dict argument, no loading code needed. We only need an explicit `dictionary=` because the Python decoders allocate a fresh buffer per call.

In regards to verification: dropping the dictionary breaks several real archives outright. Happy to share the numbers.

I am working on a test set of small, real tibx files for real world testing, based on your comments.This is coming soon

Comment (yunzheng) [reply dissect/archive/tibx/codec.py]: ah yes, I can see [LZ4_decompress_safe_continue](https://github.com/lz4/lz4/blob/0774d05537f9762f838f7ab541b7765f1a729cb5/lib/lz4.c#L1716) being used in a LSM related function. And this lz4 function internally uses a dictionary.

Comment (falklindner) [reply dissect/archive/tibx/codec.py]: Implemented in ec224b3 — dissect.util now handles every block that needs no dictionary (first of each chain, plus whole LZ4 segments). The in-tree decoder is left only for dictionary blocks when the C lz4 package isn't installed.

File: dissect/archive/tibx/c_tibx.py (hunk 1 above)
Comment (falklindner) [reply dissect/archive/tibx/c_tibx.py]: Follow-up in 9b05168: every remaining on-disk record now has a struct here (ARCH header, TLV entries, LZ4 block and cell-group headers, LDIR values, segment_map records), plus `c_boot` for the boot-sector fields used to size volumes.

File: dissect/archive/tibx/crypto.py (hunk 7 above)
Comment (falklindner) [reply dissect/archive/tibx/crypto.py]: Follow-up: the keymap's LZ4 header now goes through `c_tibx.lz4_block_header` (9b05168). 57fa473 also adds AES-GCM segment decryption, the Cyber Protect 15 default, validated on real archives.

File: dissect/archive/tibx/page.py (hunk 5 above)
Comment (falklindner) [reply dissect/archive/tibx/page.py]: `SuperBlock` now reads through the nested `arch_superblock.body` (9b05168). The agreed `commit_roots` hybrid is unchanged.

File: dissect/archive/tibx/lsm.py (hunk 4 above)
Comment (falklindner) [reply dissect/archive/tibx/lsm.py]: Follow-up in 9b05168: the ARCH header, TLV entries, cell-group headers and LDIR values now go through structs, so no hardcoded offsets are left. Attribution unchanged.

File: dissect/archive/tibx/codec.py (hunk 9 above)
Comment (falklindner) [reply dissect/archive/tibx/codec.py]: Linked-LZ4 block headers now go through `c_tibx.lz4_block_header` (9b05168). The only raw read left is the 2-byte LZ4 match offset inside the decoder, same as `dissect.util`'s.

File: tests/_synth.py (hunk 6 above)
Comment (falklindner) [reply tests/_synth.py]: The builders now use the same `c_tibx` structs as the parser (9b05168), so the fixtures can't drift from its layout. References unchanged.

File: dissect/archive/tibx/tibx.py
Comment (falklindner) [inline dissect/archive/tibx/tibx.py]: Two behaviour changes in 57fa473, both checked against acrocmd on real Cyber Protect 15 archives. `recovery_points()` now reads the slices tree, one entry per backup; counting commit roots gave too many. `volumes()` skips Acronis's own metadata streams, and `streams()` still lists everything. Older recovery points were verified file by file.

File: dissect/archive/tibx/map.py
Comment (falklindner) [inline dissect/archive/tibx/map.py]: segment_map records now go through `segment_map_key` / `segment_map_value` (9b05168). `page_count` stays the one little-endian field.

File: dissect/archive/tibx/segment.py
Comment (falklindner) [inline dissect/archive/tibx/segment.py]: AES-GCM segments (57fa473): the payload is `IV || tag || ciphertext` and is not PKCS#7-unpadded.

**Comment by falklindner** [comment]
> It would be nice to add tibx archives created by Acronis itself

Real samples are ready: 10 archives from Acronis Cyber Protect 15, covering your list (simple, password-protected, whole disk MBR/GPT vs. file-level) plus chains, stored and multi-volume. They're identity-scrubbed, about 6 MB gzipped in total. That's well above the existing `tests/_data` (534 KB), so: add them via LFS under `tests/_data/`, or would you prefer a subset?

**Comment by yunzheng** [comment]
> > It would be nice to add tibx archives created by Acronis itself
> 
> Real samples are ready: 10 archives from Acronis Cyber Protect 15, covering your list (simple, password-protected, whole disk MBR/GPT vs. file-level) plus chains, stored and multi-volume. They're identity-scrubbed, about 6 MB gzipped in total. That's well above the existing `tests/_data` (534 KB), so: add them via LFS under `tests/_data/`, or would you prefer a subset?

Nice! Can you put them in LFS under `test/_data/acronis/tibx`? The total size is fine.

**Comment by falklindner** [comment]
Added in 2cd6de9: eleven files under `tests/_data/acronis/tibx` via Git LFS (ten samples, one of them a split set). `.gitattributes` now matches `tests/_data/**` — the old `tests/_data/*` does not match subdirectories.

`tests/test_tibx_acronis.py` checks them against acrocmd's own listings: backup GUIDs per recovery point, volume count, and volume content pinned by hash.
